chandra_aca/star_probs.py

```diff
@@ -497,3 +499,49 @@ def prob_minus_p_acq(mag):
         mag = float(mag)
 
     return mag
+
+
+def guide_count(mags, t_ccd):
+    """
+    Given mags from guide stars and a temperature, calculate a guide star
+    fractional count/metric using signal-to-noise scaled mag thresholds.
+
+    :param mags: mags of guide star catalog stars
+    :param t_ccd: ACA CCD temperature at expected time of observation
+    :returns: fractional count
+    """
+    # The bright limit does not scale.
+    thresh0 = 5.9
+    thresh1 = snr_mag_for_t_ccd(t_ccd, ref_mag=10.0, ref_t_ccd=-10.9)
+    thresh2 = snr_mag_for_t_ccd(t_ccd, ref_mag=10.2, ref_t_ccd=-10.9)
+    thresh3 = snr_mag_for_t_ccd(t_ccd, ref_mag=10.3, ref_t_ccd=-10.9)
+    counts = np.zeros_like(mags)
+    counts[(mags <= thresh1) & (mags > thresh0)] = 1.0
+    counts[(mags <= thresh2) & (mags > thresh1)] = 0.75
+    counts[(mags <= thresh3) & (mags > thresh2)] = 0.5
+    return np.sum(counts)
+
+
+def t_ccd_warm_limit_for_guide(mags, min_guide_count=4.0, warm_t_ccd=-5.0, cold_t_ccd=-16.0):
+    """
+    Solve for the warmest temperature that still gets the min_guide_count.
+    This uses a brute/grid search for a reasonable maximum/warmest value.
+    This returns a value between ``cold_t_ccd`` and ``warm_t_ccd``.  At the
+    cold end the result may be below ``min_n_acq``, in which case the star catalog
+    may be rejected.
+
+    :param mags: list of star ACA mags
+    :param min_guide_count: float minimum fractional guide count
+    :param warm_t_ccd: warmest CCD temperature to consider (default=-5 C)
+    :param cold_t_ccd: coldest CCD temperature to consider (default=-16 C)
+
+    :returns: t_ccd
+    """
+    if guide_count(mags, warm_t_ccd) >= min_guide_count:
+        return warm_t_ccd
+    if guide_count(mags, cold_t_ccd) < min_guide_count:
+        return cold_t_ccd
+    def merit_func(t_ccd):
+        count = guide_count(mags, t_ccd)
+        return 0.01 + count - min_guide_count
```

Comment:
Is something like this allowed?  It looks like adding a small number to the count makes this merit function nonzero for all of the values of t_ccd where count is equal to 4 so there is then a real zero crossing at the "warm end" of that range.  This does appear to be much faster than manually computing the array of values of guide_count over some sampling of the interval.

Reply:
Excellent idea.  🥇

Reply:
But I should have added my git comment to the code so I'd have a clue about why this is there if we come back to it.